File: 02. Algorithms/01. Arrays/00. Sorting/05. Merge Sort/Merge_Sort.java
// TODO: Sort an Array using Merge Sort Technique
package array.sorting;

import java.util.Arrays;

public class Merge_Sort {

	public static void main(String[] args) {
		int arr[] = { 4, 21, 1, 9, -12, 5, 32, 2, 6 };
		System.out.println("Array before Sorting: " + Arrays.toString(arr));

		int n = arr.length;
		sort(arr, 0, n - 1);
	}

	static void sort(int arr[], int l, int r) {
		// check if there are at least 2 elements in the subarray
		if (l < r) {
			int mid = (l + r) / 2;
			sort(arr, l, mid);
			sort(arr, mid + 1, r);
			merge(arr, l, mid, r);
		}
	}

	static void merge(int arr[], int l, int mid, int r) {
		// create a temporary array to store the merged array
		int temp[] = new int[r - l + 1];

		// create pointers to the original array
		int i = l;
		int j = mid + 1;

		// create pointer to the temporary array
		int k = l;

		while (i <= mid && j <= r) {
			if (arr[i] < arr[j]) {
				temp[k] = arr[i];
				i++;
			} else {
				temp[k] = arr[j];
				j++;
			}
			k++;
		}
	}
}

File: 02. Algorithms/02. Sorting/05. Merge Sort/MergeSort.cpp
// MergeSort
#include <iostream>
using namespace std;

void merge(int *a, int *b, int *c, int s, int e) {
	int mid = (s + e) / 2;

	int i = mid;
	int j = e;
	int k = e;

	while (i >= s && j >= mid + 1) {
		if (b[i] > c[j]) {
			a[k--] = b[i--];
		}
		else {
			a[k--] = c[j--];
		}
	}
	while (i >= s) {
		a[k--] = b[i--];
	}
	while (j >= mid + 1) {
		a[k--] = c[j--];
	}
}

void MergeSort(int *a, int s, int e) {

	// base case
	if (s >= e) {
		return;
	}

	// recursive case
	// 1. Divide
	int b[100], c[100];
	int mid = (s + e) / 2;
	for (int i = s ; i <= mid ; i++) {
		b[i] = a[i];
	}
	for (int i = mid + 1; i <= e ; i++) {
		c[i] = a[i];
	}

	// 2. Sort
	MergeSort(b, s, mid);
	MergeSort(c, mid + 1, e);

	// 3. Merge
	merge(a, b, c, s, e);
}

int main() {

	int n;
	cin >> n;
	int *a = new int[n];
	for (int i = 0 ; i < n ; i++) {
		cin >> a[i];
	}

	cout << "Before Sorting: ";
	for (int i = 0 ; i < n ; i++) {
		cout << a[i] << ' ';
	}
	cout << endl;
    
	MergeSort(a, 0, n - 1);

	cout << "After Sorting: ";
	for (int i = 0 ; i < n ; i++) {
		cout << a[i] << ' ';
	}
	cout << endl;

	return 0;
}